cumulus/primitives/utility/src/lib.rs

```diff
@@ -286,23 +291,238 @@ impl<
 /// in such assetId for that amount of weight
 pub trait ChargeWeightInFungibles<AccountId, Assets: fungibles::Inspect<AccountId>> {
 	fn charge_weight_in_fungibles(
-		asset_id: <Assets as Inspect<AccountId>>::AssetId,
+		asset_id: <Assets as fungibles::Inspect<AccountId>>::AssetId,
 		weight: Weight,
-	) -> Result<<Assets as Inspect<AccountId>>::Balance, XcmError>;
+	) -> Result<<Assets as fungibles::Inspect<AccountId>>::Balance, XcmError>;
+}
+
+/// Provides an implementation of [`WeightTrader`] to charge for weight using the first asset
+/// specified in the `payment` argument.
+///
+/// The asset used to pay for the weight must differ from the `Target` asset and be exchangeable for
+/// the same `Target` asset through `SwapCredit`.
+///
+/// ### Parameters:
+/// - `Target`: the asset into which the user's payment will be exchanged using `SwapCredit`.
+/// - `SwapCredit`: mechanism used for the exchange of the user's payment asset into the `Target`.
+/// - `WeightToFee`: weight to the `Target` asset fee calculator.
+/// - `Fungibles`: registry of fungible assets.
+/// - `FungiblesAssetMatcher`: utility for mapping [`MultiAsset`] to `Fungibles::AssetId` and
+///   `Fungibles::Balance`.
+/// - `OnUnbalanced`: handler for the fee payment.
+/// - `AccountId`: the account identifier type.
+pub struct SwapFirstAssetTrader<
```

Comment:
Why have this here and not in xcm-builder? Looks like a trader other teams would like to use

Reply:
have no strong opinion here, we just have another weight trader in this crate.